ndk/src/utils.rs

```diff
@@ -12,3 +14,73 @@ pub(crate) fn status_to_io_result<T>(status: i32, value: T) -> Result<T> {
         r => unreachable!("Status is positive integer {}", r),
     }
 }
+
+pub(crate) fn android_log(level: Level, tag: &CStr, msg: &CStr) {
+    let prio = match level {
+        Level::Error => ffi::android_LogPriority::ANDROID_LOG_ERROR,
+        Level::Warn => ffi::android_LogPriority::ANDROID_LOG_WARN,
+        Level::Info => ffi::android_LogPriority::ANDROID_LOG_INFO,
+        Level::Debug => ffi::android_LogPriority::ANDROID_LOG_DEBUG,
+        Level::Trace => ffi::android_LogPriority::ANDROID_LOG_VERBOSE,
+    };
+    unsafe {
+        ffi::__android_log_write(prio.0 as c_int, tag.as_ptr(), msg.as_ptr());
+    }
+}
+
+pub(crate) fn log_panic(panic: Box<dyn std::any::Any + Send>) {
+    let rust_panic = unsafe { CStr::from_bytes_with_nul_unchecked(b"RustPanic\0") };
+
+    // Use the Rust logger if installed and enabled, otherwise fall back to the Android system
+    // logger so there is at least some record of the panic
+    let use_log = log_enabled!(Level::Error);
+
+    match panic.downcast::<String>() {
+        Ok(panic_string) => {
+            if use_log {
+                error!("RustPanic: {}", panic_string);
+            } else if let Ok(msg) = CString::new(*panic_string) {
+                android_log(Level::Error, rust_panic, &msg);
+            }
+        }
+        Err(panic) => match panic.downcast::<&str>() {
+            Ok(panic_str) => {
+                if use_log {
+                    error!("RustPanic: {}", panic_str);
+                } else if let Ok(msg) = CString::new(*panic_str) {
+                    android_log(Level::Error, rust_panic, &msg);
+                }
+            }
+            Err(_) => {
+                if use_log {
+                    error!("UnknownPanic");
+                } else {
+                    let unknown_panic =
+                        unsafe { CStr::from_bytes_with_nul_unchecked(b"UnknownPanic\0") };
+                    android_log(Level::Error, unknown_panic, unsafe {
+                        CStr::from_bytes_with_nul_unchecked(b"\0")
```

Comment:
I'd be happier with `CStr::default()`.

Reply:
`Default` is actually implemented on `&CStr`, and it's not at all clear to me how to invoke this explicitly, so I have just done it as `Default::default()`.

Reply:
Good point, I didn't think through that the returned type is `&CStr`, not `CStr`. In case you're curious, parenthesis that are normally used to adjust precedence in expressions are replaced by angle brackets when doing this to types: `<&CStr>::default()`.

Reply:
That's good to know, thanks. Want me to change it or shall I just leave it?